### PR TITLE
feat(python): ConnectionStrategy supports the pickle module

### PR DESCRIPTION
**Pull request by MarquessV**
This is to support rigetti/pyQuil#1719

### CODE REVIEW

File: crates/python/src/qpu/api.rs [hunk 1]
```diff
@@ -608,4 +611,41 @@ impl PyConnectionStrategy {
             _ => py.NotImplemented(),
         }
     }
+
+    // // Implementing this method is required for the pickle module to recognize the class as
+    // // pickleable, but __reduce__ is self sufficient, so we can just return an empty dictionary here.
+    // fn __getstate__<'py>(&'py self, py: Python<'py>) -> &'py PyDict {
+    //     PyDict::new(py)
+    // }
+    //
```
Comment (antalsz) [inline crates/python/src/qpu/api.rs:620]: Can this be deleted?

File: crates/python/tests/qpu/test_api.py [hunk 2]
```diff
@@ -44,3 +47,15 @@ def test_submit_retrieve(
 
     job_id = submit(program, memory, quantum_processor_id)
     results = retrieve_results(job_id)
+
+class TestPickle():
+    @pytest.mark.parametrize("strategy", [ConnectionStrategy.gateway(), ConnectionStrategy.direct_access(), ConnectionStrategy.endpoint_id("endpoint_id")])
+    def test_connection_strategy(self, strategy: ConnectionStrategy):
+        pickled = pickle.dumps(strategy)
+        unpickled = pickle.loads(pickled)
```
Comment (antalsz) [inline crates/python/tests/qpu/test_api.py:55]: Unless I'm missing something about pytest, this test is missing an `assert`

**Review by antalsz — CHANGES_REQUESTED** [review]
Looks like there's a missing assert in one of the tests, but the implementations look good.  Also the branch name is *incredible*.

File: crates/python/src/qpu/api.rs (hunk 1 above)
Comment (MarquessV) [reply crates/python/src/qpu/api.rs]: Sure can

File: crates/python/tests/qpu/test_api.py (hunk 2 above)
Comment (MarquessV) [reply crates/python/tests/qpu/test_api.py:55]: Oops, good catch

**Pull request merged**